### DISCUSSION TITLE
_get_descriptor

### DISCUSSION THREAD

**Discussion by BigPowerseaman** [hathach/tinyusb#1486] [Q&A]
Hello, I have a question about _get_descriptor function.
static bool _get_descriptor(uint8_t daddr, uint8_t type, uint8_t index, uint16_t language_id, void *buffer,
			    uint16_t len, tuh_xfer_cb_t complete_cb, uintptr_t user_data)
{
	tusb_control_request_t const request = {
		.bmRequestType_bit =
		{
			.recipient = TUSB_REQ_RCPT_DEVICE,
			.type      = TUSB_REQ_TYPE_STANDARD,
			.direction = TUSB_DIR_IN
		},
		.bRequest = TUSB_REQ_GET_DESCRIPTOR,
		.wValue   = tu_htole16(TU_U16(type, index)),
		.wIndex   = tu_htole16(language_id),
		.wLength  = tu_htole16(len)
	};

	tuh_xfer_t xfer = {
		.daddr       = daddr,
		.ep_addr     = 0,
		.setup       = &request,
		.buffer      = buffer,
		.complete_cb = complete_cb,
		.user_data   = user_data
	};

	bool const ret = tuh_control_xfer(&xfer);

	// if blocking, user_data could be pointed to xfer_result
	if (!complete_cb && user_data) {
		*((xfer_result_t *) user_data) = xfer.result;
	}

	return ret;
}

In the _get_descriptor function
"
if (!complete_cb && user_data) {
		*((xfer_result_t *) user_data) = xfer.result;
	}
"
Why should this code be written like this and what functions should it achieve? 
@jeremyherbert @tannewt @gpshead @bikeNomad

**Comment by hathach** [comment]
can you clarify:
- provide link to exactly where file/line of the code in question
- also format the above code properly
- is there any issues with way the code is written ?
